pandas/_libs/tslibs/strptime.pyx

```diff
@@ -129,12 +130,22 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
             if val in nat_strings:
                 iresult[i] = NPY_NAT
                 continue
+        elif checknull_with_nat_and_na(val):
+            iresult[i] = NPY_NAT
+            continue
+        elif isinstance(val, datetime):
```

Comment:
can you use PyDateTime_Check here

Reply:
probably need to cimport `import_datetime` and have a line that just says "import_datetime" (take a look at other files in this directory to see how that works)

Reply:
Thanks, that worked! Unfortunately, changing the condition breaks support for `Timestamp` objects, breaking in the same way as originally reported for `datetime.datetime` objects. When using `isisntance(val, datetime)`, `Timestamp` objects are also considered. So, I'll use `isinstance` for now

Reply:
which tests fail if you do that? I just tried applying 
```diff
diff --git a/pandas/_libs/tslibs/strptime.pyx b/pandas/_libs/tslibs/strptime.pyx
index 5ec8494f9f..4f935d51bb 100644
--- a/pandas/_libs/tslibs/strptime.pyx
+++ b/pandas/_libs/tslibs/strptime.pyx
@@ -5,6 +5,12 @@ from cpython.datetime cimport (
     datetime,
     tzinfo,
 )
+from cpython.datetime cimport (
+    PyDateTime_Check,
+    import_datetime,
+)
+
+import_datetime()
 
 from _thread import allocate_lock as _thread_allocate_lock
 
@@ -133,7 +139,7 @@ def array_strptime(ndarray[object] values, str fmt, bint exact=True, errors='rai
         elif checknull_with_nat_and_na(val):
             iresult[i] = NPY_NAT
             continue
-        elif isinstance(val, datetime):
+        elif PyDateTime_Check(val):
             val = val.strftime(fmt)
         else:
             val = str(val)
```

and then
```
pytest pandas/tests/tools/test_to_datetime.py
```
all passed

Reply:
id suggest taking a look at how we handle the `elif PyDateTime_Check(val)` case in `array_to_datetime`

Reply:
You're right @MarcoGorelli , seems like I tested while applying multiple changes instead of isolating the usage of `PyDateTime_Check`. I'll update the PR